### PR TITLE
S3C-1985: Reduce memory usage of listing algorithms

### PR DESCRIPTION
**Pull request by ghost**
The issue discovered in the field was that an key with a heavy amount of data
could lead to an exhaustion of the available memory for Node.JS javascript
code.
In the case of the buckets handling, that would be MPU objects with an
important number of parts, listed in the "location" field.

A customer case revealed and triggered this corner case, loop-crashing the
processes responsible for the listing of the database, which were using the
Arsenal.algos.list.* extensions.

That issue is hereby fixed by this commit through the trimming of the known
heavy unused data: the "location" field of the bucket's values.
Note that as the code previously didn't need to parse the value before
forwarding it back to the caller, we are now parsing it, removing the unwanted
fields, and re-stringifying it for each selected entry in the listing.

A notable impact of this change is that the CPU usage might go up substantially
while listing the contents of any bucket.
Additionally, and for safety purposes, if the data cannot be parsed and
alterated, it will be returned as-is, at the risk of maintaining the memory
consuming behavior in the case of that corner case; while an error is logged on
stderr.

Fixes S3C-1985

### CODE REVIEW

File: lib/algos/list/Extension.js [hunk 1]
```diff
@@ -23,6 +23,32 @@ class Extension {
...
+    /**
+     * Filters-out non-requested optional fields from the value. This function
+     * shall be applied on any value that is to be returned as part of the
+     * result of a listing extension.
+     *
+     * @param {Object} value - The JSON value of a listing item
+     *
+     * @return {Object} The value without the filtered-out fields.
+     */
+    filterContent(value) {
```
Comment (ghost) [inline lib/algos/list/Extension.js]: I was hesitant about adding a code comment explaining the issue here (similarly to the commit message).
I could also add a message hinting to read the commit message associated to the addition of the function.

Please tell me what you think would be best (or do the associated edit yourself, either way :) )

**Comment by ghost** [comment]
@bert-e force_reset

File: lib/algos/list/delimiter.js [hunk 2]
```diff
@@ -61,8 +61,8 @@ class Delimiter extends Extension {
      *                                                   alphabetically ordered
      *                                                   or not
      */
-    constructor(parameters) {
-        super(parameters);
+    constructor(parameters, logger) {
```
Comment (ghost) [inline lib/algos/list/delimiter.js:66]: As described in commit number 2, I had to add this for tests compatibility.
As the loggers are actually provided by cloudserver & metadata, there should be no issue here, and we should have the logger to log in the base Extension's `filterContent()` method.

**Comment by ghost** [comment]
@bert-e force_reset

File: lib/algos/list/Extension.js [hunk 3]
```diff
@@ -23,6 +23,32 @@ class Extension {
...
+        let ret = undefined;
+        try {
+            ret = JSON.parse(value);
+            delete ret.location;
+            ret = JSON.stringify(ret);
+        } catch (e) {
+            this.logger.warn('Could not parse Object Metadata while listing',
```
Comment (ghost) [inline lib/algos/list/Extension.js]: If deemed necessary, we could protect against an undefined logger here with a conditional construct.
Wasn't sure it'd be worth it, so just mentionning it here.

File: lib/algos/list/Extension.js [hunk 4]
```diff
@@ -23,6 +23,32 @@ class Extension {
...
+     * result of a listing extension.
+     *
+     * @param {Object} value - The JSON value of a listing item
+     *
+     * @return {Object} The value without the filtered-out fields.
+     */
+    filterContent(value) {
+        let ret = undefined;
+        try {
+            ret = JSON.parse(value);
```
Comment (jonathan-gramain) [inline lib/algos/list/Extension.js]: As an improvement attempt to limit the CPU cost I would prefer to have a heuristic condition on the blob size before executing this, so the location field will be trimmed only if the blob is really big. Then we pay the serialize/deserialize cost only when we're certain to gain a significant number of bytes in the result (only if our assumption that the "location" field takes up the bulk of the blob bytes, which in theory should be the case).

This should amortize the cost only to memory-greedy listings and not have an impact on the vast majority of listings, considering the blob size check to be much cheaper than the serialization cycle.

File: lib/algos/list/Extension.js (hunk 1 above)
Comment (jonathan-gramain) [inline lib/algos/list/Extension.js]: minor: renaming to `trimContents` would be more precise I think

**Review by jonathan-gramain — COMMENTED** [review]
I will attempt the changes I suggested and test on my lab.

**Review by jonathan-gramain — APPROVED** [review]
Please someone validates my changes as well

**Review by vrancurel — APPROVED** [review]
Ideally we should strip down any sub-document which is larger than a treshold, e.g. in some cases user-metadata or tags can also be > 10K, and therefore we should also strip them down.

**Comment by jonathan-gramain** [comment]
So we would have to extend the `delete` list to those potentially large attributes too when we're in the 10K+ case? Why not, just we would have to do the thorough analysis that it does not break any code that might need them (same kind than has been done for the location field already). Not sure it's worth it for now IMHO but open to discussion.

**Comment by vrancurel** [comment]
> Not sure it's worth it for now IMHO but open to discussion.

This is why I approved the PR :)

**Comment by vrancurel** [comment]
The correct way to implement this would have been acc/to me to have the listing returning only the minimal attributes needed for supporting the S3 listing (e.g. lastModified and maybe others). And do an optional params.projection API typical of document stores where consumers would typically asks for specific fields to be returned. I know it might be useful one day to be able to retrieve sub-documents while listing. I am not aware of an existing use case right now. Also we would have needed to change cloudserver+bucketclient+bucketd+repd which would have been complicated.

This is why also I approved this PR.

**Comment by jonathan-gramain** [comment]
@bert-e reset

File: lib/algos/list/Extension.js [hunk 5]
```diff
@@ -30,22 +47,28 @@ class Extension {
...
+        if (value.length >= TRIM_METADATA_MIN_BLOB_SIZE) {
+            try {
+                ret = JSON.parse(value);
+                delete ret.location;
+                ret = JSON.stringify(ret);
+            } catch (e) {
+                // Prefer returning an unfiltered data rather than
+                // stopping the service in case of parsing failure.
+                // The risk of this approach is a potential
+                // reproduction of MD-692, where too much memory is
```
Comment (ghost) [inline lib/algos/list/Extension.js:65]: Might need rewording into S3C-1985, though I think the link MD-692 should still be valid.

File: lib/algos/list/Extension.js (hunk 1 above)
Comment (ghost) [reply lib/algos/list/Extension.js]: ok for me

File: lib/algos/list/Extension.js (hunk 4 above)
Comment (ghost) [reply lib/algos/list/Extension.js]: Good point. LGTM, that was one of my worries

**Review by rahulreddy — APPROVED** [review]
LGTM. So we are not going ahead with the optional params idea?

**Comment by ghost** [comment]
@rahulreddy Indeed, going from the principle that the field which was an actual issue was never used in CloudServer, as opposed to the suspect one we were looking at (partsLocation) when devising this idea.

**Comment by ghost** [comment]
Given that both Tests on MetaData and CloudServer are green with this patch (branch bugfix/S3C-1985-runTestsWithArsenalPatch , if you're looking for it in the builds listings); I'm considering this patch as OK, given the approval of both @jonathan-gramain and @rahulreddy.

**Comment by ghost** [comment]
@bert-e approve

**Comment by ghost** [comment]
the PR was stuck on the tests of one integration branch which were not properly triggered.
I forced them, to unblock the situation.

**Pull request merged**